### PR TITLE
add OPath parser to utility

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                    case -1:
+                        OPathSegments.Add(new OPathSegment
+                        {
+                            SegmentName = segment
+                        });
+                        break;
+                    case 0:
+                        throw new ArgumentException($"There is a invalid segment {segment}");
+                    default:
+                        var keyValue = segment.Substring(index).Trim('[', ']').Split('=');
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs]: it should be throw: `abc[]def="xyz"]`

File: src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    public class OverwriteUtility
+    {
+        public static List<OPathSegment> ParseOPath(string OPathString)
+        {
+            if (string.IsNullOrEmpty(OPathString))
+            {
+                throw new ArgumentException("OPathString cannot be null or empty.", nameof(OPathString));
+            }
+
+            var segments = OPathString.Split('/');
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs]: >var segments = OPathString.Split('/'); [](start = 12, length = 38)

What value contains `/`? like `a/[id="a/b"]/c`

File: src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.OverwriteDocuments
+{
+    using System;
+    using System.Collections.Generic;
+
+    public class OverwriteUtility
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs]: >OverwriteUtility [](start = 17, length = 16)

add some ut?

File: src/Microsoft.DocAsCode.Build.OverwriteDocuments/OPathSegment.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+namespace Microsoft.DocAsCode.Build.OverwriteDocuments
+{
+    using System;
+
+    [Serializable]
+    public class OPathSegment
+    {
+        public string SegmentName { get; set; }
+
+        public string key { get; set; }
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.OverwriteDocuments/OPathSegment.cs]: >k [](start = 22, length = 1)

k => K

File: src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text.RegularExpressions;
+
+    public class OverwriteUtility
+    {
+        private static readonly Regex OPathRegex =
+            new Regex(
+                @"^(?<propertyName>[\w\.\-]+)(\[(?<key>[\w\.\-]+)=""(?<value>[\w\(\)\.\{\}\[\]\|\/@`<>: ]+)""\])?/?",
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs]: >\w\.\- [](start = 36, length = 6)

Although the property name can support any chars in YAML spec, but as we borrow the syntax from XPath, we want to limit the name as a valid **XML** name: [https://www.w3.org/TR/2008/REC-xml-20081126/#NT-NameStartChar](https://www.w3.org/TR/2008/REC-xml-20081126/#NT-NameStartChar)

Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs]: >""(?<value>[\w\(\)\.\{\}\[\]\|\/@`<>: ]+)"" [](start = 66, length = 43)

quotes is optional, if the value is int

File: src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text.RegularExpressions;
+
+    public class OverwriteUtility
+    {
+        private static readonly Regex OPathRegex =
+            new Regex(
+                @"^(?<propertyName>[:A-Za-z_][\w\.\-:]*)(\[(?<key>[:A-Za-z_][\w\.\-:]*)=((""(?<stringValue>[\w\(\)\.\{\}\[\]\|\/@ `<>:]+)"")|(?<numberValue>[-+]?[0-9]*(\.[0-9]+)?))\])?/?",
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs]: >stringValue [](start = 95, length = 11)

is boolean supported, e.g. parameters[enable=true]

Comment (928PJY) [reply src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs]: yes, it should be supported

---
In reply to: [162852574](https://github.com/dotnet/docfx/pull/2363#discussion_r162852574) [](ancestors = 162852574)

Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteUtility.cs]: ``^\s*(?<propertyName>[:A-Za-z_][\w\.\-:]*)\s*(?:\[\s*(?<key>[:A-Za-z_][\w\.\-:]*)\s*=\s*(?:(?:"(?<stringValue>[\w\(\)\.\{\}\[\]\|\/@ `<>:]+)")|(?<numberValue>[-+]?[0-9]*(\.[0-9]+)?))\s*\])?\s*(?:\/|$)``

File: test/Microsoft.DocAsCode.Build.OverwriteDocuments.Tests/OverwriteUtilityTest.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    using Xunit;
+
+    [Trait("Owner", "jipe")]
+    [Trait("EntityType", nameof(OverwriteUtility))]
+    public class OverwriteUtilityTest
+    {
+        [Fact]
+        public void ParseOPathTest()
+        {
+            var OPathstring = "a/f [c= \"d\"]/g/b[c =\"d/d_d d\"]/e";
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Build.OverwriteDocuments.Tests/OverwriteUtilityTest.cs]: >/ [](start = 32, length = 1)

is whitespace allowed around `/`?

**Pull request merged**